### ISSUE TITLE
Optimize format for short bitmaps

### ISSUE THREAD

**Issue by lemire** [RoaringBitmap/RoaringBitmap#86]
Bitmaps spanning a small space (far less than 2^16 bits) are not necessarily optimally supported right now by Roaring. This could be improved.

**Comment by ponkin** [comment]
@lemire  Does it mean that other encoders, for example [Arithmetic encoder,](https://en.wikipedia.org/wiki/Arithmetic_coding) will gain more compression rate for "sparse" bit vectors?

**Comment by lemire** [comment]
@ponkin  I think one has to set the right context. Roaring bitmaps are data structures that support set operations efficiently. If your goal is only to compress the data, but you don't really need to access the data in compressed form, then there are other alternatives that might be preferable. Please see https://github.com/lemire/JavaFastPFOR for a collection of useful compression functions on sets of integers.

**Comment by ponkin** [comment]
@lemire Thanks, that is exactly what I was lookig for. The idea is very simple: use RoaringBitmap in memory to modify it, compress bitmap during serialization phase(to send over the network or save on disk)! Thanks again!

**Comment by Kineolyan** [comment]
Hello everyone,

I'm reviving a bit this discussion.
I work for a company that is developing a in-memory database system and we are considering migrating our custom bitmap structures to RoaringBitmap. However, some of our memory benchmarks showed that RoaringBitmaps are not suited when they are scarcely filled.
Imagine two fields of a same table T, A with high cardinality of distinct members (1M), B with a few distinct members (1K). When indexing each member for all lines for both fiields, A produces a lot of bitmaps containing <=5 values, while B produces bitmaps quite filled - but not full.

For both, the computation performances are really good. However, regarding memory, we pay the cost of RoaringBitmap object overhead on small objects.
In v0.6.35, for one point, there is:
```
  RoaringBitmap header      
+ highLowContainer ref
  + RoaringArray header
  + (keys) short[4]
  + (values) Container[4]
  + (size) int
    + ArrayContainer header
    + (cardinality) int
    + (content) short[4]
```
For a 64bit JVM, we reach about 134 bytes, and 152 bytes without compressed opts.

I see that help is requested for this, so I assume it is not a priority or a goal to improve it for future versions. Yet, have you already some ideas of improvements ? Or, as important, have you tried some solutions and rejected them as they were degrading performance too much ?

Cheers

**Comment by lemire** [comment]
> I see that help is requested for this, so I assume it is not a priority or a goal to improve it for future versions.

I don't think that it is necessarily a correct interpretation. It is a community-driven project so what is a priority depends on what code people are willing to write.

**Comment by Kineolyan** [comment]
You are right, sorry for the wrong interpretation. I will see what I can do about it.
Any idea or lead to follow or exclude are welcome.

**Comment by Kineolyan** [comment]
Hello @lemire, I am working at a slow pace on an alternative to RoaringArray, providing a better management for slow values.
I reused the same principle as ArrayContainer and BitmapContainer. For a small number of values, use directly a sorted array. From a given number of values, use an optimized version.
As for containers, the main complexity is to bind properly operations between the various tasks.

Does it look good to you ?
Cheers

**Comment by lemire** [comment]
@Kineolyan  I certainly encourage experimentations.
